src/sentry/snuba/metrics_layer/api.py

```diff
@@ -0,0 +1,328 @@
+    key:value (_ key:value)?
+    in which the only supported operator is AND until this logic is switched to the metrics layer.
+    """
+    if query is None:
+        return None
+
+    filters = []
+    matches = QUERY_REGEX.findall(query)
+    for key, value in matches:
+        filters.append(Filter(key=key, value=value))
+
+    if not query:
+        raise InvalidMetricsQuery("Error while parsing the query.")
+
+    return filters
+
+
+def _parse_group_by(group_bys: Optional[Sequence[str]]) -> Optional[Sequence[GroupBy]]:
+    """
+    This function supports parsing in the form:
+    value (_ value)?
+    """
+    if group_bys is None:
+        return None
+
+    # For uniformity, we also convert the group bys to internal dataclasses.
+    return [GroupBy(key=group_by) for group_by in group_bys]
+
+
+def _build_snql_query(
+    field: Field,
+    snql_filters: Optional[ConditionGroup],
+    snql_group_bys: Optional[List[Union[Column, AliasedExpression]]],
+) -> Timeseries:
+    if field.refers_to_mri():
+        metric = Metric(mri=field.metric_name)
+    else:
+        metric = Metric(public_name=field.metric_name)
+
+    return Timeseries(
+        metric=metric,
+        aggregate=field.aggregate,
+        filters=snql_filters,
+        groupby=snql_group_bys,
+    )
+
+
+def _filters_to_snql(filters: Optional[Sequence[Filter]]) -> Optional[ConditionGroup]:
+    if filters is None:
+        return None
+
+    conditions = []
+    for _filter in filters:
+        condition = Condition(lhs=Column(name=_filter.key), op=Op.EQ, rhs=_filter.value)
+        conditions.append(condition)
+
+    return conditions
+
+
+def _group_bys_to_snql(
+    group_bys: Optional[Sequence[GroupBy]],
+) -> Optional[List[Union[Column, AliasedExpression]]]:
+    if group_bys is None:
+        return None
+
+    columns = []
+
+    for group_by in group_bys:
+        columns.append(Column(name=group_by.key))
+
+    return columns
+
+
+def _get_granularity(interval: int) -> int:
+    best_granularity: Optional[int] = None
+
+    for granularity in sorted(GRANULARITIES):
+        if granularity <= interval:
+            best_granularity = granularity
+
+    if best_granularity is None:
+        raise InvalidMetricsQuery("The interval specified is lower than the minimum granularity")
+
+    return best_granularity
+
+
+def _build_intervals(start: datetime, end: datetime, interval: int) -> Sequence[datetime]:
+    """
+    Builds a list of all the intervals that are queried by the metrics layer.
+    """
+    start_seconds = start.timestamp()
+    end_seconds = end.timestamp()
+
+    current_time = start_seconds
+    intervals = [datetime.fromtimestamp(current_time, timezone.utc)]
+    while current_time + interval < end_seconds:
+        next_time = current_time + interval
+        intervals.append(datetime.fromtimestamp(next_time, timezone.utc))
+        current_time = next_time
+
+    return intervals
+
+
+def _zerofill_series(
```

Comment:
Is this all that the zerofill needed to do? I imagined it would build a row with all the same keys, but a zero value. E.g., A query that wants a row for time 1, 2, 3, but there is no data for time 2:

```
[
   {"time": 1, "aggregate_value": 65},
   {"time": 3, "aggregate_value": 45},
]
```

Note a missing row for `"time": 2`. What I thought the zerofill would do is add a proper row value, so the final result would be:

```
[
   {"time": 1, "aggregate_value": 65},
   {"time": 2, "aggregate_value": 0},    <- new row
   {"time": 3, "aggregate_value": 45},
]
```

Reply:
Yep, initially it was filling zero elements but now I put `None` since this was the old API behavior from my testing. I will have to change the name of the function.

This algorithm does zerofilling in the way you said, but in the inverse, it builds all the intervals that are requested and defaults them to `None` (but can easily be `0`) and then for each returned value it substitutes the `None` with the actual value.